src/gc.c

```diff
@@ -3385,6 +3366,14 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
         if (sweep_full)
             gc_sweep_perm_alloc();
     }
+
+    size_t heap_size = jl_atomic_load_relaxed(&gc_heap_stats.heap_size);
+    if (heap_size > max_total_memory*0.8)
+        next_sweep_full = 1;
+    else
+        next_sweep_full = 0;
+    size_t new_heap_target = 2 * heap_size > max_total_memory ? max_total_memory : 2 * heap_size;
```

Comment:
unconditionally doubling the heap size seems wrong, right?

Reply:
It's after collection, the idea is to let you double your live set before doing a GC. This is not uncommon

Reply:
That seems much too fast as you get close to max heap. You only get about 4-5 doublings before running out of memory entirely on 32-bit. For something more nuanced: https://github.com/JuliaLang/julia/blob/ba0e4843cb67eeea6882ddabf30c8c26bf22e1f1/src/array.c#L735-L752

Reply:
So this is only doubling live heap which means in the case you described we are on a purely allocating load and don't free anything, go uses basically this (+ GC assist stuff that doesn't apply). However I'm implementing https://dl.acm.org/doi/pdf/10.1145/3563323 which claims to be a more correct thing. The implementation is based on https://github.com/mmtk/mmtk-core/blob/40c977308cb24d8bdf46b82916a388c10fa956e7/src/util/heap/gc_trigger.rs#L205